### ISSUE TITLE
How to get the Solution Pool using Julia/CPLEX?

### ISSUE THREAD

**Issue by GiAbreu** [jump-dev/CPLEX.jl#197]
Hi everyone,

I need the solution pool for storing multiple solutions of a MIP model. I using Julia and the packages JuMP and CPLEX as a solver. So I need to collect solutions within a given percentage of the optimal solution but I don’t know how to do it in Julia, anybody can help me, please?

**Comment by IssamT** [comment]
I don't think it's available on the JuMP layer since it's solver specific. If you are ready to do a bit of hacking, then you can maybe try using `setincumbentcallbackfunc` of CPLEX.jl

**Comment by GiAbreu** [comment]
I set the parameters as explained on the IBM website, but I don't know how to call the populate solution pool. Using callbacks I only got the values ​of the objective function, best bound, numbers of nodes of B&B tree, but I need the values ​​of the multiple solutions for the variables that have the same objective function value, which means populate the solution pool.

**Comment by odow** [comment]
We now wrap `CPXpopulate(model.env, model.lp)`, but there is still no nice way of accessing multiple solutions via MOI. Ideally, we would have something like the following:
```Julia
model = CPLEX.Optimizer()
x = MOI.add_variable(model)
# ... model definition ...
MOI.optimize!(model)
CPXpopulate(model.env, model.lp)
for n = 1:MOI.get(model, MOI.ResultCount())
    x_n = MOI.get(model, MOI.VariablePrimal(n), x)
    println("Value of x in solution $(n) is $(x_n)")
end
```
This is all possible via the exposed C API, so it just needs some plumbing. I won't be implementing this, but I will leave the issue open for a motivated person in the community.

**Comment by davide-f** [comment]
Hi, are there updates on this issue?
Thank you

**Comment by odow** [comment]
@jd-foster worked out the details here: https://github.com/jump-dev/jump-dev.github.io/pull/58. It just needs implementing at the MOI level. PRs accepted.

**Comment by odow** [comment]
Closing this as won't-fix. It has been a few years and there doesn't seem to be much demand for this. It also requires some explicit C API calls, rather than the option setting like Gurobi which makes it a bit more annoying to implement.

@jd-foster has written a tutorial that demonstrates how to get multiple solutions with CPLEX:
See https://jump.dev/tutorials/2021/11/02/tutorial-multi-jdf/

If you are wanting multiple solutions with CPLEX and that isn't sufficient, please comment below and we can consider re-opening this issue.

**Comment by chriscoey** [comment]
@odow this was closed a few months back, so I imagine the story hasn't changed since then. Do you have a sense for how much work it would be to expose this in a more user-friendly way here? Even better would be if we could have better support in MOI for getting multiple solutions from MIP solvers.

**Comment by odow** [comment]
The issue is that CPLEX doesn't neatly provide multiple solutions. You need to create an populate a separate solution pool object, etc.

MOI has great support for querying multiple solutions. See, e.g, Gurobi. The complexity is on the solver side, not JuMP or MOI.

I would consider PRs to CPLEX.jl, but this is not a priority for me.

**Comment by chriscoey** [comment]
Thanks, makes sense. I guess by suggesting something at the MOI level I meant more around the concept of solution pools, but you are saying that just the concept of multiple solutions is enough for MOI, as evidenced by Gurobi.jl.

**Comment by odow** [comment]
> as evidenced by Gurobi.jl

And also MiniZinc.jl, MultiObjectiveAlgorithms.jl, etc.

I don't want to add a solution pool concept to MOI. The solution here is to add some sort of special option to CPLEX.jl that, instead of solving normally, solves via the solution pool and makes the set of solutions available. It's perfectly do-able, but it requires some amount of work.

**Comment by odow** [comment]
Re-opening as maybe-planned

**Comment by chriscoey** [comment]
Based on this answer https://github.com/ERGO-Code/HiGHS/issues/1193#issuecomment-2941630152 do you think something analogous could be achieved for HiGHS.jl?

**Comment by jajhall** [comment]
> Based on this answer [ERGO-Code/HiGHS#1193 (comment)](https://github.com/ERGO-Code/HiGHS/issues/1193#issuecomment-2941630152) do you think something analogous could be achieved for HiGHS.jl?

Maybe you need to use the callback

**Comment by odow** [comment]
I don't have plans to add this to HiGHS. But I re-opened this issue because @hdavid16 sent me an email about funding it, so perhaps we could expand the scope to include HiGHS.

My main comment would be: the HiGHS callback approach would only return the solutions that were incidentally found on the way to the optimal solution. It wouldn't have support to enumerate a range of near-optimal solutions. Is that what you want?

The callback example is https://github.com/jump-dev/MiniZinc.jl/issues/95#issuecomment-2917892636

**Comment by jajhall** [comment]
> My main comment would be: the HiGHS callback approach would only return the solutions that were incidentally found on the way to the optimal solution. It wouldn't have support to enumerate a range of near-optimal solutions. Is that what you want?

Indeed, all you get are whatever integer-feasible solutions HiGHS finds when solving the MIP. 

Finding a pool of (near-)optimal solutions isn't something that we'll implement for a long time, as it's non-trivial to do so

**Comment by chriscoey** [comment]
Ah OK, thanks both - I was hoping that we'd be able to get multiple (near optimal) solutions out of HiGHS, similar to what we are discussing here with CPLEX. 

The ability to get integer feasible solutions found along the way is incidentally still useful for the particular problem that motivates @hdavid16 currently though.

**Comment by jajhall** [comment]
Maybe it wouldn't be so hard to do, but we'll parallelise the tree search first @odow

**Comment by odow** [comment]
I don't think this should be a priority for HiGHS development. There are many other more important things.
